R-package/R/lgb.Dataset.R

```diff
@@ -533,22 +533,47 @@ Dataset <- R6::R6Class(
 
     # Update parameters
     update_params = function(params) {
-
-      # Parameter updating
-      if (!lgb.is.null.handle(private$handle)) {
-        lgb.call(
-          "LGBM_DatasetUpdateParam_R"
-          , ret = NULL
-          , private$handle
+      if (lgb.is.null.handle(private$handle)) {
+        private$params <- modifyList(private$params, params)
+      } else {
+        call_state <- .Call(
+          "LGBM_DatasetUpdateParamWarning_R"
+          , lgb.params2str(private$params)
           , lgb.params2str(params)
+          , call_state
+          , PACKAGE = "lib_lightgbm"
         )
-        return(invisible(self))
+        call_state <- as.integer(call_state)
+        if (call_state != 0L) {
+
+          # raise error if raw data is freed
+          if (is.null(private$raw_data)) {
+            lgb.last_error()
+          }
+
+          # Overwrite paramms
+          private$params <- modifyList(private$params, params)
+          self$finalize()
+        }
       }
-      private$params <- modifyList(private$params, params)
       return(invisible(self))
 
     },
 
+    get_params = function() {
+      keys <- c("has_header", "header", "max_bin", "max_bin_by_feature", "bin_construct_sample_cnt", "subsample_for_bin", "min_data_in_bin", "use_missing",
```

Comment:
what is the purpose of using this list instead of just returning `private$params` directly?

Reply:
There are some failed tests in python side with returning private$params only. 

this is to write the params from Dataset to Booster. As parameters may is changes in Booster side, we only return the params of Dataset actually used, otherwise it may write some old parameters back.